Add `with_proxy_ca_certificate` and `with_proxy_excludes`

gordonwang0 opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #4713

# Rationale for this change
 
Currently, object_store has a `with_proxy_url` function that only accepts a URL. The underlying reqwest library supports more proxy options, such as proxy authentication and a 'no proxy' list, but these cannot be set in object_store.

# What changes are included in this PR?

Adds public API functions `with_proxy_ca_certificate` and `with_proxy_excludes` to allow a proxy CA certificate and proxy exclude list to be set.

# Are there any user-facing changes?

New API functions `with_proxy_ca_certificate` and `with_proxy_excludes`. The currently available `with_proxy_url` remains unchanged.

tustvold commented:
Typically we have tried to avoid exposing reqwest in public APIs, with the vague idea that we might be able to alter the underlying impl at a later date.

What do you think about exposing `with_proxy_auth`  and `with_proxy_excludes` or something?

gordonwang0 commented:
I added `with_proxy_excludes` to set a list of hosts to exclude.

I also added a function `with_proxy_ca_certificate` that sets a new CA certificate to trust. Although AWS, Azure, and GCP use publicly-rooted server certificates, some proxy environments use a MITM proxy that intercepts TLS requests. This function allows the proxy's CA certificate to be trusted.

I was able to put the username/password in the URL (e.g. `http://user:password@proxy.com`), so I didn't add `with_proxy_auth`.

tustvold commented on object_store/src/aws/mod.rs:
```diff
@@ -946,6 +946,23 @@ impl AmazonS3Builder {
     }
 
+    /// Set a trusted proxy CA certificate
+    pub fn with_proxy_ca_certificate(
```
What do you think about making these methods on ClientConfig instead of duplicated per builder?

gordonwang0 replied:
I added them to the builder because the existing `with_proxy_url` was part of the builder. I think it makes sense to keep all proxy options on the same struct, but I don't have a strong opinion on this. Do you think it's better to move it to ClientConfig?

tustvold commented:
Thank you
